Adds crc32c codec

normanrz opened the pull request:
Adds Crc32c codec, which appends a crc32c checksum to the input buffer when encoding and validates the checksum when decoding.

Fixes #610

TODO:

- [x] Unit tests and/or doctests in docstrings
- [x] Tests pass locally
- [x] Docstrings and API docs for any new/modified user-facing classes and functions
- [x] Changes documented in docs/release.rst
- [x] Docs build locally
- [x] GitHub Actions CI passes
- [x] Test coverage to 100% (Codecov passes)

dstansby commented:
Should this live in https://github.com/zarr-developers/numcodecs/blob/main/numcodecs/checksum32.py, where we already have `CRC32`? As I understand (from [a stackoverflow post](https://stackoverflow.com/questions/26429360/crc32-vs-crc32c)) `CRC32` and `CRC32C` are basically the same algorithm, so it makes sense to put them in the same namespace.

normanrz commented:
I didn't look for `checksum32.py`. Thanks for telling me.
I just looked into that. A problem is that the existing checksum codecs prepend the checksum whereas zarr3 expects the checksum to be appended. Either we single out crc32c to append the checksum or we add an option to all checksum codecs. Wdyt?

normanrz commented:
I moved the code to checksum32.py and refactored the tests (they had no parameterization) in https://github.com/zarr-developers/numcodecs/pull/613/commits/7e78d936de24c63023c592ba18ba8c576393c8d2

dstansby commented on docs/checksum32.rst:
```diff
@@ -13,6 +13,17 @@ CRC32
 
+CRC32C
+------
+.. autoclass:: CRC32C
```
😱 I forgot we were still doing things manually in the docs, this is a good reminder for me to fix this in another PR  😄

dstansby commented on numcodecs/checksum32.py:
```diff
@@ -1,39 +1,75 @@
 
 class CRC32(Checksum32):
     codec_id = 'crc32'
     checksum = zlib.crc32
```
Is it worth explicitly adding `location = 'start'` here?

dstansby commented on numcodecs/checksum32.py:
```diff
@@ -1,39 +1,75 @@
 class Checksum32(Codec):
     # override in sub-class
     checksum = None
+    location: CHECKSUM_LOCATION = 'start'
```
Would be good to add a quick docstring in `__init__` to describe the new parameter and attribute.

normanrz replied:
I added it to the class docstring. Unfortunately not very DRY. Not sure if docstrings can be dynamic?

dstansby commented on pyproject.toml:
```diff
@@ -16,6 +16,7 @@ for use in data storage and communication applications."""
 readme =  "README.rst"
 dependencies = [
     "numpy>=1.7",
+    "crc32c>=2.7",
```
I don't have any strong feelings either way, but is it worth making this an optional extra instead of being in the main dependencies?

normanrz replied:
I made it an optional dep. Seems like making it a core dependency later is easier than the other way around.

dstansby reviewed (approved):
Looks good to me - just a few questions, but none of them blocking.

normanrz commented:
I don't see why this PR isn't ready to merge (it might be just late)? Do I need another approval?

dstansby reviewed (approved):
Maybe just needs a fresh approval?
